src/mesh.cpp

```diff
@@ -102,6 +121,70 @@ StructuredMesh::bin_label(int bin) const {
   }
 }
 
+//==============================================================================
+// Unstructured Mesh implementation
+//==============================================================================
+
+UnstructuredMesh::UnstructuredMesh(pugi::xml_node node) : Mesh(node) {
+  n_dimension_ = 3;
```

Comment:
Should there be any additional checks either here, or in the derived classes, that the mesh is indeed 3D? For instance, the constructor for `libMesh::Mesh` that is used later takes a dimension; from the libMesh documentation:

> Constructor. Takes dim, the dimension of the mesh. The mesh dimension can be changed (and may automatically be changed by mesh generation/loading) later.

Does anything break if we pass 3 as the mesh dimension into `libMesh` mesh, when in fact the mesh is not 3D?

Reply:
Ah yeah we should definitely check that. I'll add that in.

Reply:
I suppose a check that the input geometry is 3D in the appropriate location would also be good.

Reply:
Yes, both of those would be good. I think you'll need to put the check on the libMesh mesh's dimension (i.e. what is in the exodus file) after calling `initialize()` based on the doc string from libMesh above.

libMesh also has two different ideas of "dimension" - for instance, a 2-D plane that isn't aligned with a Cartesian axis could be considered either 2-D or 3-D (because one coordinate x, y, or z isn't fixed for all points). I think we probably want the 2-D check.

Reply:
I was just looking at that as well. I think I used the wrong one in my latest commit. I'll make that change shortly.

Reply:
Can confirm that this check works for a 2-D plane not parallel to a Cartesian axis.

Reply:
Oh actually, but it looks like if you error at this stage, some memory isn't handled correctly:

```
 Reading tallies XML file...
 ERROR: Mesh file mesh.e specified for use in an unstructured mesh is not a 3D
        mesh.
application called MPI_Abort(MPI_COMM_WORLD, -1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=-1
:
system msg for write_line failure : Bad file descriptor
Memory leak detected!
Compile in DEBUG mode with --enable-reference-counting
for more information
```

Maybe this isn't important since you're failing anyways, but if it could be fixed that'd be preferable for me.

Reply:
Yeah same. We should handle that differently. Thanks for checking that! I'll look into that shortly.

Reply:
Welp, I was hoping this would be simple but, apparently, all existing `libMesh::Mesh` instances need to be cleaned up before the `LibMeshInit` object we maintain gets destroyed or we'll see this output. This makes it trick as we could potentially have several mesh instances loaded and it seems messy to have to go find all of those in our vector of meshes -- particularly because all of this is happening inside an object constructor. I'll have to think on this a bit more.